diff --git a/logger.go b/logger.go
index 535ed9b..a0105a4 100644
--- a/logger.go
+++ b/logger.go
@@ -35,6 +35,8 @@ var (
 
 	// timeNow is a customizable for testing purposes.
 	timeNow = time.Now
+
+	Modules []string
 )
 
 // Record represents a log record and contains the timestamp when the record
@@ -120,6 +122,7 @@ func MustGetLogger(module string) *Logger {
 	if err != nil {
 		panic("logger: " + module + ": " + err.Error())
 	}
+	Modules = append(Modules, module)
 	return logger
 }
 
