diff --git a/vendor/github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql/sql.go b/vendor/github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql/sql.go
index 9ff6637..62deda1 100644
--- a/vendor/github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql/sql.go
+++ b/vendor/github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql/sql.go
@@ -73,6 +73,9 @@ func Open(driverName, dataSourceName string, opts ...Option) (*sql.DB, error) {
 	opts = append([]Option{regOpt}, opts...)
 	d := newDriver(db.Driver(), newTraceConfig(opts...))
 
+	if err = db.Close(); err != nil {
+	        return nil, err
+	}
 	// Use the instrumented driver to open a connection to the database.
 	if driverCtx, ok := d.(driver.DriverContext); ok {
 		connector, err := driverCtx.OpenConnector(dataSourceName)
