diff --git a/clickhouse.go b/clickhouse.go
index f1d6d4e..efd8cb9 100644
--- a/clickhouse.go
+++ b/clickhouse.go
@@ -117,7 +117,7 @@ type txOptions struct {
 }
 
 func (ch *clickhouse) beginTx(ctx context.Context, opts txOptions) (*clickhouse, error) {
-	ch.logf("[begin] tx=%t, data=%t", ch.inTransaction, ch.block != nil)
+	// ch.logf("[begin] tx=%t, data=%t", ch.inTransaction, ch.block != nil)
 	switch {
 	case ch.inTransaction:
 		return nil, sql.ErrTxDone
